### ISSUE TITLE
StackingCVClassifier

### ISSUE THREAD

**Issue by elejke** [rasbt/mlxtend#29]
Hello. 
Don't you think, that your realization of stacking, where you are fitting 1-level clfs on whole X_train and after that just predict labels(or probs) by them on the same X_train is not really good? It may lead to big overfitting of this predicted labels.
I think, that better approach is to make folds, and train clfs smth like this ( you train on (1 - 1/n_folds) part of train data set, then predict labels for the rest of X_train and do it for all folds, when you finish it, you have labels for the whole X_train, but without overfitting, and now you may fit one more time every clf to predict X_test labels (here you can use the whole X_train for fitting them)): 
```python
def fit(self, X_train, y_train):
        self.n_examples = X_train.shape[0]
        self.n_classes = len(set(y_train))
        self.folds = KFold(n=self.n_examples, n_folds=self.n_folds)

        clfs_preds = np.array([]).reshape(self.n_examples, 0)

        for clf in self.clfs:
            clf_ = clone(clf)
            clf_preds = np.array([]).reshape(0, self.n_classes*self.probas + 1 * (not self.probas))
            for train, pred in self.folds:
                clf_.fit(X_train[train], y_train[train])
                if not self.probas:
                    clf_pred = clf_.predict(X_train[pred]).reshape(X_train[pred].shape[0], 1)
                else:
                    clf_pred = clf_.predict_proba(X_train[pred]).reshape(X_train[pred].shape[0], self.n_classes)
                clf_preds = np.concatenate([clf_preds, clf_pred], axis=0)

            clfs_preds = np.concatenate([clfs_preds, clf_preds], axis=1)
        # fitting the clfs to predict X_test labels or probabilities when fitting the meta-classifier
        for clf in self.clfs:
            clf.fit(X_train, y_train)

        # add clfs predictions to X_train table or not
        if self.append_preds:
            X_train_ext = np.concatenate([X_train, clfs_preds], axis=1)
        else:
            X_train_ext = clfs_preds
        self.meta_clf.fit(X_train_ext, y_train)
        return self

    def predict(self, X_test):

        clfs_preds = np.array([]).reshape(X_test.shape[0], 0)
        for clf in self.clfs:
            if not self.probas:
                clf_pred = clf.predict(X_test).reshape(X_test.shape[0], 1)
            else:
                clf_pred = clf.predict_proba(X_test).reshape(X_test.shape[0], self.n_classes)
            clfs_preds = np.concatenate([clfs_preds, clf_pred], axis=1)

        # add clfs predictions to X_test table or not
        if self.append_preds:
            X_test_ext = np.concatenate([X_test, clfs_preds], axis=1)
        else:
            X_test_ext = clfs_preds

        return self.meta_clf.predict(X_test_ext)
```

**Comment by rasbt** [comment]
Thanks for the suggestion @elejke . You are absolutely right regarding overfitting. This was more like the implementation of the classic/original algorithm, and I had planned to follow up with the k-fold Cross-validation variant as StackingCVClassifier, e.g., as summarized in Aggarwal, Charu C., ed. Data classification: algorithms and applications. CRC Press, 2014.

https://books.google.com/books?id=nwQZCwAAQBAJ&lpg=PA500&dq=stacking%20classifier%20subsets&pg=PA499#v=onepage&q&f=false

Since you already did the work, do you want to submit it as a pull request some time? Otherwise, I will surely get to it in the next couple of weeks! Thanks!

**Comment by rasbt** [comment]
PS: The algorithm I implemented 19.7, and the CV variant you mentioned would be 19.8 1-2 pages down.

**Comment by elejke** [comment]
Of course, I can submit this code and help you to maintain it. I did it for me.
But, there are many questions in implementation. And the main problem, that it obviously works ~ n_folds times longer.
If you want, we can discuss and after that I will submit.

**Comment by reiinakano** [comment]
Hi. Perhaps another suggestion I can make to elejke's code above is the use of a Stratified Fold CV instead of a regular K Fold CV. This way, the fitting will still work properly in case the training data set is not balanced. Also, if the training data set hasn't been shuffled when it is passed into the classifier (e.g. 500 Category 1 followed by 500 Category 2 items), we risk training the base models using just one category while attempting to predict another category it has never seen before. Using Stratified Fold CV would solve both of these problems. 

Of course, there is the matter of rearranging the labels in the proper order when we train the second level classifier (since Stratified K Fold CV will very likely cause a shuffling of the predictions) but that shouldn't be too difficult to implement.

**Comment by rasbt** [comment]
Thanks for the reminder ... totally missed to follow up on this! And I totally agree with you, we should definitely use `StratifiedKFold` for the actual implementation. I am not a big fan of non-stratified k-fold in general even if datasets are balanced, since it will screw up the class proportions, especially if you have small- to medium-sized datasets. 

> since Stratified K Fold CV will very likely cause a shuffling of the predictions

conveniently, `StratifiedKFold` has a shuffle parameter :).

**Comment by reiinakano** [comment]
If no one is assigned to this yet, I'd be glad to submit it as a pull request. I'm looking for my first contribution to an open source project and this seems simple enough for me to do. :)

**Comment by rasbt** [comment]
No one is working on a PR so far -- as far as I know ;). I had this on my to do list but due to other priorities, I never got to this ... 
I would be very happy about a pull-request!

Btw. we could make the stratification optional with default to `True`. I.e., `StackingCVClassifier(..., stratify=True, ...)`

**Comment by reiinakano** [comment]
Great, I'll start working on it then.

Regarding the use_probas parameter, it seems you and elejke have different ways of implementing it. You take the average of the class probabiliites of each classifier while elejke simply appends them together and uses the class probabilities of each classifier as extra features for training the second classifier. 

I'm leaning towards implementing it as he did since we never know what information might get lost if we take the average of the class probabilities. What do you think?

**Comment by rasbt** [comment]
> 'm leaning towards implementing it as elejke did since we never know what information might get lost if we take the average of the class probabilities. What do you think?

I agree with you, I think that's what they also did in

- Ting, Kai Ming, and Ian H. Witten. "[Issues in stacked generalization.](http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.34.948)" J. Artif. Intell. Res.(JAIR) 10 (1999): 271-289.

when they observed the better performance vs. using class labels as outputs of the first-level classifiers.

I think the line

```python
clf_.predict_proba(X_train[pred]).reshape(X_train[pred].shape[0], self.n_classes)
```

should be redundant though, right?

So, just for reference, the proposed algorithm should be the following one:

![screen shot 2016-09-18 at 2 05 25 pm](https://cloud.githubusercontent.com/assets/5618407/18618030/99c26726-7dab-11e6-9a70-92398b8b9c6c.png)

**Comment by rasbt** [comment]
The StackingCVClassifier has just been added in PR #100 . Thanks @reiinakano !
